scene/3d/mesh_instance.cpp

```diff
@@ -163,15 +172,242 @@ void MeshInstance::_resolve_skeleton_path() {
+		const int index_count = mesh_instance->surface_get_array_index_len(surface_index);
+
+		uint32_t array_offsets[Mesh::ARRAY_MAX];
+		uint32_t stride = visual_server->mesh_surface_make_offsets_from_format(format, vertex_count, index_count, array_offsets);
+		uint32_t offset_vertices = array_offsets[Mesh::ARRAY_VERTEX];
+		uint32_t offset_normals = array_offsets[Mesh::ARRAY_NORMAL];
+		uint32_t offset_bones = array_offsets[Mesh::ARRAY_BONES];
+		uint32_t offset_weights = array_offsets[Mesh::ARRAY_WEIGHTS];
+
+		PoolByteArray buffer = visual_server->mesh_surface_get_array(mesh_rid, surface_index);
+		PoolByteArray::Write buffer_write = buffer.write();
+
+		for (int vertex_index = 0; vertex_index < vertex_count; ++vertex_index) {
+			float bone_weights[4];
+			if (format & Mesh::ARRAY_COMPRESS_WEIGHTS) {
+				const uint16_t *weight_ptr = (const uint16_t *)(buffer_write.ptr() + offset_weights + (vertex_index * stride));
+				bone_weights[0] = (weight_ptr[0] / (float)0xFFFF);
+				bone_weights[1] = (weight_ptr[1] / (float)0xFFFF);
+				bone_weights[2] = (weight_ptr[2] / (float)0xFFFF);
+				bone_weights[3] = (weight_ptr[3] / (float)0xFFFF);
+			} else {
+				const float *weight_ptr = (const float *)(buffer_write.ptr() + offset_weights + (vertex_index * stride));
+				bone_weights[0] = weight_ptr[0];
+				bone_weights[1] = weight_ptr[1];
+				bone_weights[2] = weight_ptr[2];
+				bone_weights[3] = weight_ptr[3];
+			}
+
+			int bone_ids[4];
+			if (format & Mesh::ARRAY_FLAG_USE_16_BIT_BONES) {
+				const uint16_t *bones_ptr = (const uint16_t *)(buffer_write.ptr() + offset_bones + (vertex_index * stride));
+				bone_ids[0] = bones_ptr[0];
+				bone_ids[1] = bones_ptr[1];
+				bone_ids[2] = bones_ptr[2];
+				bone_ids[3] = bones_ptr[3];
+			} else {
+				const uint8_t *bones_ptr = buffer_write.ptr() + offset_bones + (vertex_index * stride);
+				bone_ids[0] = bones_ptr[0];
+				bone_ids[1] = bones_ptr[1];
+				bone_ids[2] = bones_ptr[2];
+				bone_ids[3] = bones_ptr[3];
+			}
+
+			int b0 = bone_ids[0];
+			int b1 = bone_ids[1];
+			int b2 = bone_ids[2];
+			int b3 = bone_ids[3];
+
+			Transform transform;
+			transform.origin =
+					bone_weights[0] * bone_transforms[b0].origin +
+					bone_weights[1] * bone_transforms[b1].origin +
+					bone_weights[2] * bone_transforms[b2].origin +
+					bone_weights[3] * bone_transforms[b3].origin;
+
+			transform.basis =
+					bone_transforms[b0].basis * bone_weights[0] +
+					bone_transforms[b1].basis * bone_weights[1] +
+					bone_transforms[b2].basis * bone_weights[2] +
+					bone_transforms[b3].basis * bone_weights[3];
+
+			Vector3 &vertex = (Vector3 &)buffer_write[vertex_index * stride + offset_vertices];
+			vertex = transform.xform(vertex);
+
+			switch (software_skinning_normal_mode) {
+				case DONT_TRANSFORM_NORMALS: {
+					// Nothing to do.
+				} break;
+				case ENSURE_CORRECT_NORMALS: {
+					transform.basis.invert();
+					transform.basis.transpose();
```

Comment:
I think you forgot to transform the normal here.

Reply:
There's no break in the switch so the transform code is in common with the other case, but instead I can copy the two lines over if it helps with readability.

Reply:
Ah yep good point, maybe just expand on the comment for fools like myself! :+1:

Reply:
Actually I should probably duplicate the lines instead, because it's easy to miss and only saves a few lines of code.